### PR TITLE
1 implement basic content service

### PR DESCRIPTION
**Pull request by bergdola**
## Description of changes
Basic implementation of the content service under the constraint that the relation content->tag is 1:n.
  
## How has this been tested?

- [ ] automated unit test
  - TODO: we will do that in the near future 
- [x] automated integration test
- [ ] automated acceptance test
- [ ] manual, exploratory test
    
## Checklist before requesting a review
- [x] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My code fulfills all acceptance criteria
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have made corresponding changes to the documentation
- [ ] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+@Transactional
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java:20]: I don't have this annotation is other services. Why is it needed? If only needed in a specific method, consider only having it there. #Nitpick

File: src/main/java/de/unistuttgart/iste/gits/content_service/persistence/dao/ContentEntity.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    private UUID id;
+
+    @Column(nullable = false, length = 255)
+    private String name;
+
+    @Column(nullable = false)
+    private int rewardPoints;
+
+    @Column(nullable = false)
+    private boolean workedOn;
```
[the remark this answers is not in the document]
Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/persistence/dao/ContentEntity.java:30]: No todo for this pull request

**Review by PaulBredl — CHANGES_REQUESTED** [review]
As we kinda moved this card anywhere to this next sprint I was nitpicky, but if you really just want to finish the card just merge it I guess 😂

File: src/test/java/de/unistuttgart/iste/gits/content_service/ContentServiceApiTests.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+@GraphQlApiTest
+public class ContentServiceApiTests {
+    @Test
+    void shouldAddContentAndQueryBack(GraphQlTester graphQlTester) {
+        graphQlTester.documentName("create-content")
```
[the remark this answers is not in the document]
Comment (bergdola) [reply src/test/java/de/unistuttgart/iste/gits/content_service/ContentServiceApiTests.java:20]: I can understand your point. In this case, I prefer the separate file as I am re-using some queries and the name should give a clear idea, what the query is about. Additionally, IntelliJ offers code completion for this file type.

File: src/main/resources/graphql/schema.graphqls [hunk 4]
```diff
@@ -1,16 +1,87 @@
...
+    id: UUID!
+    # new name of the content, max 255 characters
+    name: String! @NotBlank @Size(max: 255)
+    # new reward points of the content
+    rewardPoints: Int!
+    # new value for "workedOn" field of the content
+    workedOn: Boolean!
+}
+
+type Tag {
```
[the remark this answers is not in the document]
Comment (bergdola) [reply src/main/resources/graphql/schema.graphqls:40]: Thank you for your feedback, content:tag relation is going to be changed to m:n in the next PR.

File: src/main/java/de/unistuttgart/iste/gits/content_service/persistence/repository/ContentRepository.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.util.List;
+import java.util.UUID;
+
+@Repository
+public interface ContentRepository extends JpaRepository<ContentEntity, UUID> {
+
+    List<ContentEntity> findByName(String name);
+
+    @Query("select t.content from Tag t where t.name = :tag")
```
[the remark this answers is not in the document]
Comment (bergdola) [reply src/main/java/de/unistuttgart/iste/gits/content_service/persistence/repository/ContentRepository.java:17]: For me, it is less a question about sufficiency but more about readability. I often prefer manually defined queries because it is very clear what happens. Do you think a comment would really help in this case?

File: src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java (hunk 1 above)
Comment (bergdola) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java:20]: Transactional is needed when a service is accessing different entities, e.g. tag and content. Otherwise you get some obscure hibernate exception. In the content microservice, we plan to offer an API where tags are a list of string attributes for a content object, hence the ContentService class will have to deal with content and tag entities. For simplicity reasons, I would like to keep the transactional annotation on class level.

File: src/test/java/de/unistuttgart/iste/gits/content_service/ContentServiceApiTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulBredl) [reply src/test/java/de/unistuttgart/iste/gits/content_service/ContentServiceApiTests.java:20]: FYI you can also get code completion in the test itself if you "show context actions" (right click or alt+einf) and "inject language" (or similar wording) and choose graphql. But it has to be done for every test.

File: src/main/java/de/unistuttgart/iste/gits/content_service/persistence/repository/ContentRepository.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/persistence/repository/ContentRepository.java:17]: I see your point. I'm just a bit careful with SQL. If something goes wrong, it's another thing that might be debugged.

File: build.gradle [hunk 6]
```diff
@@ -53,6 +73,7 @@ dependencies {
 	implementation 'org.modelmapper:modelmapper:3.+'
 	implementation 'com.graphql-java:graphql-java-extended-scalars:20.0'
 	implementation 'com.graphql-java:graphql-java-extended-validation:20.0'
+	testImplementation 'junit:junit:4.13.1'
```
Comment (PaulBredl) [inline build.gradle:76]: Stupid question, is this the old junit 4 or is this just the version number?

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          distribution: 'temurin'
+      - name: Validate Gradle wrapper
+        uses: gradle/wrapper-validation-action@ccb4328a959376b642e027874838f60f8e596de3
+      - name: Run chmod to make gradlew executable
+        run:  chmod +x gradlew
+      - name: Build with Gradle
+        uses: gradle/gradle-build-action@749f47bda3e44aa060e82d7b3ef7e40d953bd629
+        with:
+          arguments: build
+      - name: Execute tests
```
Comment (PaulBredl) [inline .github/workflows/test.yaml:28]: I think I messaged you on discord about this. Gradle build already executes tests (therefore cleanTest is required here). That's just a minor thing but we obviously don't need to run them twice. And if a test fails, we don't get a jacoco report because the build fails. Maybe just skip the build step or something? I don't know. Doesn't really matter too much anyway

**Review by PaulBredl — APPROVED** [review]
I have some questions left but other than that approved 😊
